expertsystem/state/conservation_rules.py

```diff
@@ -342,279 +175,241 @@
+            )
         return True
 
 
-@rule_conditions(
-    variable_conditions=[
-        (StateQuantumNumberNames.Parity, [DefinedForAllEdges()]),
-        (ParticlePropertyNames.Pid, [DefinedForAllOutgoingEdges()]),
-        (StateQuantumNumberNames.Spin, [DefinedForAllOutgoingEdges()]),
-    ]
-)
+@dataclass(frozen=True)
+class IdenticalParticleSymmetryEdgeInput:
+    parity: EdgeQuantumNumbers.parity
+    spin_magnitude: EdgeQuantumNumbers.spin_magnitude
+    spin_projection: EdgeQuantumNumbers.spin_projection
+    pid: EdgeQuantumNumbers.pid
+
+
 class IdenticalParticleSymmetrization(Rule):
-    def __call__(self, ingoing_part_qns, outgoing_part_qns, _):
+    def __call__(  # type: ignore
+        self,
+        ingoing_edge_qns: List[IdenticalParticleSymmetryEdgeInput],
+        outgoing_edge_qns: List[IdenticalParticleSymmetryEdgeInput],
+        _=None,
+    ) -> bool:
         """Implementation of particle symmetrization."""
 
-        def _check_particles_identical(particles):
+        def _check_particles_identical(
+            particles: List[IdenticalParticleSymmetryEdgeInput],
+        ) -> bool:
             """Check if pids and spins match."""
-            pid_label = ParticlePropertyNames.Pid
-            spin_label = StateQuantumNumberNames.Spin
-            reference_pid = particles[0][pid_label]
-            reference_spin = particles[0][spin_label]
+            reference_pid = particles[0].pid
+            reference_spin_proj = particles[0].spin_projection
             for particle in particles[1:]:
-                if particle[pid_label] != reference_pid:
+                if particle.pid != reference_pid:
                     return False
-                if particle[spin_label] != reference_spin:
+                if particle.spin_projection != reference_spin_proj:
                     return False
             return True
 
-        if _check_particles_identical(outgoing_part_qns):
-            parity_label = StateQuantumNumberNames.Parity
-
-            if is_boson(outgoing_part_qns[0]):
+        if _check_particles_identical(outgoing_edge_qns):
+            if is_boson(outgoing_edge_qns[0].spin_magnitude):
                 # we have a boson, check if parity of mother is even
-                parity = ingoing_part_qns[0][parity_label]
+                parity = ingoing_edge_qns[0].parity
                 if parity == -1:
                     # if its odd then return False
                     return False
             else:
                 # its fermion
-                parity = ingoing_part_qns[0][parity_label]
+                parity = ingoing_edge_qns[0].parity
                 if parity == 1:
                     return False
 
         return True
 
 
-def _is_clebsch_gordan_coefficient_zero(spin1, spin2, spin_coupled):
+def _is_clebsch_gordan_coefficient_zero(
+    spin1: Spin, spin2: Spin, spin_coupled: Spin
+) -> bool:
     m_1 = spin1.projection
     j_1 = spin1.magnitude
     m_2 = spin2.projection
```

Comment:
Is that more pythonic?

Reply:
Mwa, just a bit less clutter:
```python
    is_zero = False
    if (j_1 == j_2 and m_1 == m_2) or (m_1 == 0.0 and m_2 == 0.0):
        if abs(mag - j_1 - j_2) % 2 == 1:
            is_zero = True
    elif j_1 == mag and m_1 == -proj:
        if abs(j_2 - j_1 - mag) % 2 == 1:
            is_zero = True
    elif j_2 == mag and m_2 == -proj:
        if abs(j_1 - j_2 - mag) % 2 == 1:
            is_zero = True
    return is_zero
```
vs
```python
    if (j_1 == j_2 and m_1 == m_2) or (m_1 == 0.0 and m_2 == 0.0):
        if abs(mag - j_1 - j_2) % 2 == 1:
            return True
    if j_1 == mag and m_1 == -proj:
        if abs(j_2 - j_1 - mag) % 2 == 1:
            return True
    if j_2 == mag and m_2 == -proj:
        if abs(j_1 - j_2 - mag) % 2 == 1:
            return True
    return False
```
I've added changed it in a fork PR

Reply:
I guess here its really up to preference. Btw the reason why i did this, is because of a "good practice" rule that functions with a single return statement are a bit easier to follow